HIP buffer transfer - Eliminate cupy usage

swetha097 opened the pull request:
Support for Hip buffer transfer by eliminate the usage of cupy

rrawther commented on rocAL/rocAL/source/rali_api_data_transfer.cpp:
```diff
@@ -111,3 +111,48 @@ raliCopyToOutput(
+    try
+    {
+        auto tensor_layout = (tensor_format == RALI_NHWC) ?  RaliTensorFormat::NHWC : RaliTensorFormat::NCHW;
+        context->master_graph->copy_hip_out_tensor(out_ptr, tensor_layout, multiplier0, multiplier1, multiplier2,
```
use the copy_out_tensor from rocAL with device pointer

rrawther commented on rocAL/rocAL_pybind/amd/rali/pipeline.py:
```diff
@@ -240,22 +241,36 @@ def copyImage(self, array):
 
+    def copyToHipTensorNHWC(self, array, multiplier, offset, reverse_channels, tensor_dtype):
+        if tensor_dtype == types.FLOAT:
+            b.raliCopyToHipOutputTensor32(self._handle, ctypes.c_void_p(array.data_ptr()), types.NHWC,
```
make changes as per the review comments for c++ api

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/rali_pybind.cpp:
```diff
@@ -35,6 +35,18 @@ namespace pybind11
 namespace rali{
     using namespace pybind11::literals; // NOLINT
     // PYBIND11_MODULE(rali_backend_impl, m) {
+        static void *ctypes_void_ptr(const py::object &object)
```
adjust spacing here

swetha097 replied:
Done

swetha097 commented on rocAL/rocAL/include/rali_api_data_transfer.h:
```diff
@@ -43,4 +43,15 @@ extern "C"  RaliStatus   RALI_API_CALL raliCopyToOutputTensor16(RaliContext rali
                                                               float offset1, float offset2,
                                                               bool reverse_channels);
+
+extern "C" RaliStatus RALI_API_CALL raliCopyToHipOutputTensor32(RaliContext p_context, void *out_ptr, RaliTensorLayout tensor_format,
```
Done

swetha097 commented:
> reviewed with comments

@rrawther 
Addressed the review comments
Confirmed by running the host backend & hip backend trainings in the MI100 Machine

rrawther reviewed (approved):
@LakshmiKumar23 : please verify classification and SSD trainings with these changes

kiritigowda commented:
@rrawther - CI pipeline enabled - http://math-ci.rocm.amd.com/blue/organizations/jenkins/compute-rocm-dkms-no-npi-hipclang%2FShort-GPUOpen%2FMIVisionX/detail/PR-765/1/pipeline

@LakshmiKumar23 let us know if this looks good with your test.

LakshmiKumar23 commented:
> @LakshmiKumar23 : please verify classification and SSD trainings with these changes

@rrawther Does this need to be run on any particular hardware specifically?

kiritigowda commented:
@rrawther rerunning CI due to failure. @LakshmiKumar23 can you confirm this PR for merge.

LakshmiKumar23 commented:
do

> @rrawther rerunning CI due to failure. @LakshmiKumar23 can you confirm this PR for merge.

@kiritigowda I am trying to test it now

kiritigowda commented:
@LakshmiKumar23 any updates on this?

rrawther reviewed (commented):
@swetha097 : There are few codacy errors. Can you fix those so we can merge this PR.
Also it seems to have merge conflicts with tot

kiritigowda commented:
> @kiritigowda : Can you merge this PR. Swetha has confirmed it is working

waiting for CI to finish -> http://math-ci.rocm.amd.com/blue/organizations/jenkins/compute-rocm-dkms-no-npi-hipclang%2FShort-GPUOpen%2FMIVisionX/detail/PR-765/4/pipeline
